cmd/gpud/command/join.go

```diff
@@ -0,0 +1,297 @@
+package command
+
+import (
+	"archive/tar"
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/urfave/cli"
+
+	"github.com/leptonai/gpud/components/accelerator"
+	"github.com/leptonai/gpud/components/state"
+	"github.com/leptonai/gpud/config"
+	"github.com/leptonai/gpud/internal/login"
+	"github.com/leptonai/gpud/log"
+	latency_edge "github.com/leptonai/gpud/pkg/latency/edge"
+	"github.com/leptonai/gpud/pkg/process"
+)
+
+func cmdJoin(cliContext *cli.Context) (retErr error) {
+	rootCtx, rootCancel := context.WithTimeout(context.Background(), 3*time.Minute)
+	defer rootCancel()
+	endpoint := cliContext.String("endpoint")
+	clusterName := cliContext.String("cluster-name")
+	provider := cliContext.String("provider")
+	nodeGroup := cliContext.String("node-group")
+	extraInfo := cliContext.String("extra-info")
+
+	stateFile, err := config.DefaultStateFile()
+	if err != nil {
+		return fmt.Errorf("failed to get state file: %w", err)
+	}
+	db, err := state.Open(stateFile)
+	if err != nil {
+		return fmt.Errorf("failed to open state file: %w", err)
+	}
+	defer db.Close()
+
+	uid, _, err := state.CreateMachineIDIfNotExist(rootCtx, db, "")
+	if err != nil {
+		return fmt.Errorf("failed to get machine uid: %w", err)
+	}
+
+	cmd := exec.Command("nproc")
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	if err = cmd.Run(); err != nil {
+		return fmt.Errorf("executing nproc: %w", err)
+	}
+
+	totalCPU, err := strconv.ParseInt(strings.TrimSpace(out.String()), 10, 64)
+	if err != nil {
+		return fmt.Errorf("error parsing cpu: %w", err)
+	}
+
+	_, productName, err := accelerator.DetectTypeAndProductName(rootCtx)
+	if err != nil {
+		return err
+	}
+
+	// network section
+	publicIP, _ := login.PublicIP()
+	region := "unknown"
+	detectProvider := ""
```

Comment:
Remove this? Currently hard-coded to tailscale only since we only talk to tailscale public DERP servers:
 https://github.com/leptonai/gpud/blob/871533e7efcff60e1b1addeed7dd8c94d9845bbc/pkg/latency/edge/derp.go#L73-L74

(this has nothing to do with node setup or GPUs)